cloud/lib/api-stack.ts

```diff
@@ -0,0 +1,189 @@
+	PropagatedTagSource,
+	Secret as EnvSecret,
+	TaskDefinition,
+} from 'aws-cdk-lib/aws-ecs';
+import {
+	ApplicationLoadBalancer,
+	ApplicationProtocol,
+	ListenerAction,
+	ListenerCondition,
+} from 'aws-cdk-lib/aws-elasticloadbalancingv2';
+import { ARecord, IHostedZone, RecordTarget } from 'aws-cdk-lib/aws-route53';
+import { LoadBalancerTarget } from 'aws-cdk-lib/aws-route53-targets';
+import { Bucket } from 'aws-cdk-lib/aws-s3';
+import { Secret } from 'aws-cdk-lib/aws-secretsmanager';
+import { RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib/core';
+import { Construct } from 'constructs';
+import { join } from 'node:path';
+
+import { appName, resourceId, stageName } from './resourceNamingUtils';
+
+type ApiStackProps = StackProps & {
+	apiDomainName: string;
+	certificate: ICertificate;
+	customAuthHeaderName: string;
+	customAuthHeaderValue: string;
+	domainName: string;
+	hostedZone: IHostedZone;
+};
+
+export class ApiStack extends Stack {
+	constructor(scope: Construct, id: string, props: ApiStackProps) {
+		super(scope, id, props);
+
+		const generateResourceId = resourceId(scope);
+		const stage = stageName(scope);
+
+		const {
+			apiDomainName,
+			certificate,
+			customAuthHeaderName,
+			customAuthHeaderValue,
+			domainName,
+			env,
+			hostedZone,
+		} = props;
+		const region = env?.region;
+		if (!region) {
+			throw new Error('Region not defined in stack env, cannot continue!');
+		}
+
+		const dockerImageAsset = new DockerImageAsset(this, generateResourceId('container-image'), {
+			directory: join(__dirname, '../../backend/'),
+		});
+
+		// TODO Look into IPv6 Routing, so no need for NAT Gateway or Instance!
+		// https://www.turbogeek.co.uk/aws-cdk-ipv6-v/
+		// https://docs.aws.amazon.com/vpc/latest/userguide/egress-only-internet-gateway.html
+
+		// AMI courtesy of fck-nat: https://fck-nat.dev/stable/deploying/#cdk
+		const natGatewayProvider = NatInstanceProviderV2.instanceV2({
+			instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
+			machineImage: new LookupMachineImage({
+				name: 'fck-nat-al2023-*-x86_64-ebs',
+				owners: ['568608671756'],
+			}),
+			defaultAllowedTraffic: NatTrafficDirection.NONE,
+		});
+
+		const vpc = new Vpc(this, generateResourceId('vpc'), {
+			maxAzs: 2,
+			natGatewayProvider,
+		});
+
+		// Security Group rules for NAT instance: from container to internet
+		natGatewayProvider.connections.allowToAnyIpv4(Port.HTTPS, 'Egress from NAT to internet');
+		vpc.privateSubnets.forEach((subnet) => {
+			natGatewayProvider.connections.allowFrom(
+				Peer.ipv4(subnet.ipv4CidrBlock),
+				Port.HTTPS,
+				`Ingress from private subnet ${subnet.subnetId} to NAT`
+			);
+		});
+
+		const apiKeySecret = Secret.fromSecretNameV2(
+			this,
+			generateResourceId('apikey'),
+			`${stage}/SpyLogic/ApiKey`
+		);
+
+		/*
+		Fargate service defines task running our backend container.
+			Note: cannot use ApplicationLoadBalancedFargateService construct as that adds a
```

Comment:
I'm not sure but you might be able to use the [`openListener`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ecs_patterns.ApplicationLoadBalancedFargateService.html#openlistenerl) property on the `ApplicationLoadBalancedFargateService` construct to make its listener private. But the approach below works as well.

Reply:
Yeah that didn't work nicely for me... One problem is around the Security Groups that construct generates, but additionally it creates a default rule on the listener that you cannot remove, and we want our default rule to deny access. So, I needed to do things the hard way 🙄